appengine/standard/firebase/firetactoe/firetactoe.py

```diff
@@ -0,0 +1,270 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tic Tac Toe with the Firebase API"""
+
+import datetime
+from functools import wraps
+import json
+import os
+import re
+import urllib
+
+from Crypto.PublicKey import RSA
+import flask
+from flask import request
+from google.appengine.api import users
+from google.appengine.ext import db
+import jwt
+import requests
+import requests_toolbelt.adapters.appengine
+
+# Use the App Engine Requests adapter. This makes sure that Requests uses
+# URLFetch.
+requests_toolbelt.adapters.appengine.monkeypatch()
+
+
+# Firebase database server authorization key
+# This is used to authenticate the server and authorize write access
+#
+# Replace this value with your Firebase db key from:
+#
+#     https://console.firebase.google.com/project/_/settings/database
+#
+FIRE_DB_KEY = 'REPLACE_WITH_YOUR_DB_KEY'
+
+_FIREBASE_CONFIG_TEMPLATE = '_firebase_config.html'
+_SERVICE_ACCOUNT_FILENAME = 'credentials.json'
+
+_CWD = os.path.dirname(__file__)
+_IDENTITY_ENDPOINT = ('https://identitytoolkit.googleapis.com/'
+                      'google.identity.identitytoolkit.v1.IdentityToolkit')
+# Firebase database url/channels memoization
+_FIRE_URL = None
+
+
+app = flask.Flask(__name__)
+
+
+def _get_firebase_config():
+    """Parses the client-side Firebase config into a dict."""
+    with open(os.path.join(_CWD, 'templates', _FIREBASE_CONFIG_TEMPLATE)) as f:
+        html = f.read()
+    obj = re.search(r'{([^}]+)}', html, re.MULTILINE)
+    params = [p.strip().split(':', 1) for p in obj.group(1).split(',')]
+    return dict((k.strip('" '), v.strip('" ')) for k, v in params)
+
+
+def _send_firebase_message(u_id, message=None):
+    global _FIRE_URL
```

Comment:
That's fine, but can you pull all of this caching stuff into it's own function? `_get_fire_url`?